pkg/collector/corechecks/gpu/gpu.go

```diff
@@ -0,0 +1,203 @@
+		}
+	}
+
+	if m.timeResolver == nil {
+		m.timeResolver, err = sectime.NewResolver()
+		if err != nil {
+			return fmt.Errorf("cannot create time resolver: %s", err)
+		}
+	}
+	return nil
+}
+
+// ensureProcessor ensures that there is a stats processor for the given key
+func (m *Check) ensureProcessor(key *model.StreamKey, snd sender.Sender, gpuThreads int, checkDuration time.Duration) {
+	if _, ok := m.statProcessors[key.Pid]; !ok {
+		m.statProcessors[key.Pid] = &StatsProcessor{
+			key: key,
+		}
+	}
+
+	m.statProcessors[key.Pid].totalThreadSecondsUsed = 0
+	m.statProcessors[key.Pid].sender = snd
+	m.statProcessors[key.Pid].gpuMaxThreads = gpuThreads
+	m.statProcessors[key.Pid].measuredInterval = checkDuration
+	m.statProcessors[key.Pid].timeResolver = m.timeResolver
+	m.statProcessors[key.Pid].lastCheck = m.lastCheckTime
+}
+
+// Run executes the check
+func (m *Check) Run() error {
+	if err := m.ensureInitialized(); err != nil {
+		return err
+	}
+
+	gpuDevices, err := getGPUDevices()
+	if err != nil {
+		return fmt.Errorf("get GPU devices: %s", err)
+	}
+
+	data, err := m.sysProbeUtil.GetCheck(sysconfig.GPUMonitoringModule)
+	if err != nil {
+		return fmt.Errorf("get gpu check: %s", err)
+	}
+	now := time.Now()
+
+	var checkDuration time.Duration
+	// mark the check duration as close to the actual check as possible
+	if !m.lastCheckTime.IsZero() {
+		checkDuration = now.Sub(m.lastCheckTime)
+	}
+
+	snd, err := m.GetSender()
+	if err != nil {
+		return fmt.Errorf("get metric sender: %s", err)
+	}
+
+	stats, ok := data.(model.GPUStats)
+	if !ok {
+		return log.Errorf("ebpf check raw data has incorrect type: %T", stats)
+	}
+
+	// TODO: Multiple GPUs are not supported yet
+	gpuThreads, err := gpuDevices[0].GetMaxThreads()
+	if err != nil {
+		return fmt.Errorf("get GPU device threads: %s", err)
+	}
+
+	usedProcessors := make(map[uint32]bool)
+
+	for _, data := range stats.CurrentData {
+		m.ensureProcessor(&data.Key, snd, gpuThreads, checkDuration)
+		m.statProcessors[data.Key.Pid].processCurrentData(data)
+		usedProcessors[data.Key.Pid] = true
+	}
+
+	for _, data := range stats.PastData {
+		m.ensureProcessor(&data.Key, snd, gpuThreads, checkDuration)
+		m.statProcessors[data.Key.Pid].processPastData(data)
+		usedProcessors[data.Key.Pid] = true
+	}
+
+	// As we compute the utilization based on the number of threads launched by the kernel, we need to
+	// normalize the utlization if we get above 100%, as the GPU can enqueue threads.
+	totalGPUUtilization := 0.0
+	for _, processor := range m.statProcessors {
+		if usedProcessors[processor.key.Pid] {
+			totalGPUUtilization += processor.getGPUUtilization()
+		}
+	}
+	normFactor := max(1.0, totalGPUUtilization)
+
+	for _, processor := range m.statProcessors {
+		if usedProcessors[processor.key.Pid] {
+			processor.setGPUUtilizationNormalizationFactor(normFactor)
+			err := processor.markInterval()
+			if err != nil {
+				return fmt.Errorf("mark interval: %s", err)
```

Comment:
Do you want to return an error without calling `sender.Commit()`? You've maybe got pending statistics queued up.

Reply:
I'm not completely sure. The errors from `markInterval` can only come from the `sender.GaugeWithTimestamp` method, and that seems to produce an error only if the timestamp is invalid. I think in that case it's better to no t send stats and just report an error, but I'm not sure what's the usual practice in this case.

Reply:
I think the problem is that the stats will get sent on next run? I don't see a way to discard any queued stats.

Reply:
Changed to have the `snd.Commit()` call with a `defer` statement